readme.md

[![Build Status](https://travis-ci.org/MatthewDaws/TileMapBase.svg?branch=master)](https://travis-ci.org/MatthewDaws/TileMapBase) 

# TileMapBase

Uses OpenStreetMap tiles, or other tile servers, to produce "basemaps" for use with matplotlib.  Uses a SQLite database to cache the tiles, so you can experiment with map production without re-downloading the same tiles.  Supports Open Data tiles from the
UK Ordnance Survey.


## Requirements

Pure python.  Uses [requests](http://docs.python-requests.org/en/master/) to make HTTP requests for tiles, and [pillow](https://python-pillow.org/) for image manipulation.


## Install

    pip install tilemapbase

or build from source:

    python setup.py install

or directly from GitHub:

    pip install https://github.com/MatthewDaws/TileMapBase/zipball/master


## Example

- [Example](https://github.com/MatthewDaws/TileMapBase/blob/master/notebooks/Example.ipynb) - Jupyter notebook showing examples.
- [Ordnance Survey](https://github.com/MatthewDaws/TileMapBase/blob/master/notebooks/Ordnance%20Survey.ipynb) - Ordnance survey examples.
- [Notebooks](https://github.com/MatthewDaws/TileMapBase/blob/master/notebooks/) - Other examples.


## OpenStreetMap data

OpenStreetMap Data is "© OpenStreetMap contributors”, see http://www.openstreetmap.org/copyright

Please remember that tile set usage is subject to constraints: https://operations.osmfoundation.org/policies/tiles/

- As of 25/05/2019  [OSM requires a user agent for all requests](https://operations.osmfoundation.org/policies/tiles/), with a warning "Faking another app’s User-Agent WILL get you blocked."  We hence default to using "TileMapBase" as a user agent.

## Ordnance Survery data

[Contains OS data © Crown copyright and database right (2017)](http://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/)


changes
# 0.4.7

- Stop noisy logging, see https://github.com/MatthewDaws/TileMapBase/issues/13
- Parse the new geoPandas CRS code, see https://github.com/MatthewDaws/TileMapBase/issues/19
- Some error handling around lon/lat, see https://github.com/MatthewDaws/TileMapBase/issues/23

# 0.4.6

- Stop caching invalid responses from remote server: https://github.com/MatthewDaws/TileMapBase/pull/16
  This fixes https://github.com/MatthewDaws/TileMapBase/issues/3

# 0.4.5

- OSM now needs headers: https://github.com/MatthewDaws/TileMapBase/pull/10
- https://github.com/MatthewDaws/TileMapBase/issues/11


# 0.4.4

- Add option to `to_expect` method to allow enlarging the rectangle.


# 0.4.3

- Add Extent static constructors from epsg:3857 coords
- Work to allow uploading to PyPi:
    - Download and install `pandoc`
    - `pip install pypandoc`
    - Delete the `dist` directory
    - `python setup.py sdist`
    - `pip install twine`
    -   The following assumes ".pypirc" has been written correctly (see https://packaging.python.org/specifications/pypirc/)
    - `twine upload --repository pypitest dist/*`
    -   If you get an error, maybe I have already uploaded, and then deleted, the same version.  Try changing the version by adding letters.
    - Test: ``
    - More instructions here: https://packaging.python.org/tutorials/packaging-projects/

# 0.4.2

- Allow SQLite cache to operate across multiple threads.


# 0.4.1

- Round tile coordinates correctly in `Plotter`


# 0.4.0

- Further ONS support


# 0.3.0

- Working Ordnance Survey tile support
- More unit tests.
- Improved cache support


# 0.2.0

- More object oriented and immutable interface in "mapping.py"

tests/test_cache.py
import pytest
import unittest.mock as mock
import datetime, sqlite3, os, threading

import tilemapbase.cache as cache

class CacheTest(cache.ConcreteCache):
    def __init__(self):
        self.get = None
        self.gets = []
        self.placed = []

    def get_from_cache(self, str_request):
        self.gets.append(str_request)
        return self.get

    def place_in_cache(self, str_request, obj_as_bytes):
        self.placed.append((str_request, obj_as_bytes))

@pytest.fixture()
def cache_test():
    executor = mock.MagicMock()
    concrete_cache = mock.MagicMock()
    c = cache.Cache(executor, concrete_cache)
    obj_expected = bytes([1,2,12,18])
    executor.fetch.return_value = obj_expected
    concrete_cache.get_from_cache.return_value = None
    return c, executor, concrete_cache, obj_expected

def test_Cache_notInCache_requests(cache_test):
    c, executor, _, _ = cache_test
    c.fetch("spam")
    executor.fetch.assert_called_with("spam")

def test_Cache_notInCache_places(cache_test):
    c, executor, ccache, expected_obj = cache_test
    obj = c.fetch("spam")
    assert( obj == expected_obj )
    ccache.place_in_cache.assert_called_with("spam", expected_obj)

def test_Cache_inCacheDoesntRequest(cache_test):
    c, executor, ccache, expected_obj = cache_test
    ccache.get_from_cache.return_value = (expected_obj, None)
    obj = c.fetch("spam")
    assert( obj == expected_obj )
    assert( executor.fetch.call_count == 0 )

def test_Cache_withTimeout(cache_test):
    c, executor, ccache, expected_obj = cache_test
    c.expire_time = datetime.timedelta(days=1)
    now = datetime.datetime(2016,4,10,12,30)
    ccache.get_from_cache.return_value = (expected_obj, now)
    
    with mock.patch("datetime.datetime") as datetime_mock:
        datetime_mock.now.return_value = now
        obj = c.fetch("spam")
        assert( executor.fetch.call_count == 0 )
        assert( obj == expected_obj )

def test_Cache_withTimeout_doesExpire(cache_test):
    c, executor, ccache, expected_obj = cache_test
    c.expire_time = datetime.timedelta(days=1)
    now = datetime.datetime(2016,4,10,12,30)
    ccache.get_from_cache.return_value = (expected_obj, now)
    
    with mock.patch("datetime.datetime") as datetime_mock:
        datetime_mock.now.return_value = now + datetime.timedelta(days=1, minutes=1)
        obj = c.fetch("spam")
        executor.fetch.assert_called_with("spam")
        ccache.place_in_cache.assert_called_with("spam", expected_obj)


@pytest.fixture
def db_cache():
    try:
        c = cache.SQLiteCache("test.db")
        try:
            yield c
        finally:
            c.close()
    finally:
        try:
            os.remove("test.db")
        except Exception:
            pass

def test_database_exists():
    try:
        assert( cache.database_exists("test.db") == False )

        sqlite3.connect("test.db")
        assert( cache.database_exists("test.db") == False )

        conn = sqlite3.connect("test.db")
        conn.execute("CREATE table cache (name)")
        conn.execute("CREATE table other (thing)")
        conn.close()
        assert( cache.database_exists("test.db") == True )
    finally:
        os.remove("test.db")

def test_sqcache_emplace(db_cache):
    assert(db_cache.get_from_cache("spam") is None)

    now = datetime.datetime(2016,4,10,12,30)
    strftime = datetime.datetime.strftime
    strptime = datetime.datetime.strptime
    with mock.patch("datetime.datetime") as datetime_mock:
        datetime_mock.now.return_value = now
        datetime_mock.strftime = strftime
        datetime_mock.strptime = strptime
        db_cache.place_in_cache("spam", b"eggs")
        db_cache.place_in_cache("spam1", b"eggs1")
        assert(db_cache.get_from_cache("spam") == (b"eggs", now))
        assert(db_cache.get_from_cache("spam1") == (b"eggs1", now))

def test_sqcache_query(db_cache):
    assert db_cache.query() == []

    now = datetime.datetime.now()
    db_cache.place_in_cache("spam", b"eggs")
    
    q = db_cache.query()
    assert len(q) == 1
    assert q[0][0] == "spam"
    assert abs((q[0][1] - now).total_seconds()) < 1

    db_cache.place_in_cache("spam1", b"eggs")
    q = db_cache.query()
    assert len(q) == 2
    assert set(name for name, _ in q) == {"spam", "spam1"}

def test_sqcache_update(db_cache):
    db_cache.place_in_cache("spam", b"eggs")
    q = db_cache.query()
    assert len(q) == 1
    assert q[0][0] == "spam"

    db_cache.place_in_cache("spam", b"eggs")
    q = db_cache.query()
    assert len(q) == 1
    assert q[0][0] == "spam"

def test_sqcache_remove(db_cache):
    db_cache.place_in_cache("spam", b"eggs")
    db_cache.place_in_cache("spam1", b"eggs")
    assert len(db_cache.query()) == 2

    db_cache.remove("spam")
    q = db_cache.query()
    assert len(q) == 1
    assert q[0][0] == "spam1"

def test_sqcache_multi_threading(db_cache):
    db_cache.place_in_cache("spam", b"eggs")

    barrier = threading.Barrier(2)
    def task():
        db_cache.remove("spam")
        barrier.wait()

    threading.Thread(target=task).start()
    barrier.wait(timeout=5)

    assert len(db_cache.query()) == 0


tests/test_tiles.py
import pytest
import unittest.mock as mock
import os, shutil
import PIL.Image
import datetime

import tilemapbase.tiles as tiles

@mock.patch("tilemapbase.cache.database_exists")
@mock.patch("tilemapbase.cache.SQLiteCache")
def test_init(cache, dbexists):
    tiles._sqcache = None
    tiles.init()

    assert( dbexists.called )
    name = dbexists.call_args[0][0]
    cache.assert_called_with(name)

@mock.patch("tilemapbase.cache.database_exists")
@mock.patch("tilemapbase.cache.SQLiteCache")
def test_init_throws(cache, dbexists):
    dbexists.return_value = False
    tiles._sqcache = None
    with pytest.raises(Exception):
        tiles.init()
    with pytest.raises(Exception):
        tiles._get_cache()

@pytest.fixture
def image():
    filename = os.path.join("notebooks", "test.jpg")
    with open(filename, "br") as f:
        return f.read()

import collections
Response = collections.namedtuple("Response", ["ok", "content"])

@mock.patch("tilemapbase.tiles._sqcache")
@mock.patch("requests.get")
def test_Tiles(get, sqcache, image):
    sqcache.get_from_cache.return_value = None
    get.return_value = Response(True, image)

    t = tiles.Tiles("example{zoom}/{x}/{y}.jpg", "TEST")
    x = t.get_tile(10,20,5)

    assert(x.width == 256)
    assert(x.height == 256)
    assert(get.call_args[0][0] == "example5/10/20.jpg")
    assert(sqcache.place_in_cache.call_args[0][0] == "TEST#10#20#5")

@mock.patch("tilemapbase.tiles._sqcache")
@mock.patch("requests.get")
def test_invalid_Tiles(get, sqcache):
    sqcache.get_from_cache.return_value = None
    get.return_value = Response(True, "abcdef")

    t = tiles.Tiles("example{zoom}/{x}/{y}.jpg", "TEST")
    with pytest.raises(IOError) as exc_info:
        x = t.get_tile(10,20,5)

    assert("Received invalid tile" in str(exc_info.value))

@mock.patch("tilemapbase.tiles._sqcache")
@mock.patch("requests.get")
def test_OSM(get, sqcache, image):
    sqcache.get_from_cache.return_value = None
    get.return_value = Response(True, image)

    tiles.build_OSM().get_tile(5,10,20)

    assert(get.call_args[0][0] == "http://a.tile.openstreetmap.org/20/5/10.png")

def test_Cache():
    cache_mock = mock.Mock()
    tile_cache_test = tiles.Cache(cache_mock)
    assert tile_cache_test.make_request_string("SPAM",1,2,3) == "SPAM#1#2#3"
    assert tile_cache_test.split_request_string("SPAM#5#6#77") == ("SPAM", 5, 6, 77)

    data = tile_cache_test.get_from_cache(("SPAM", 6, 3, 12))
    cache_mock.get_from_cache.assert_called_with("SPAM#6#3#12")
    assert data is cache_mock.get_from_cache.return_value

    with pytest.raises(NotImplementedError):
        tile_cache_test.place_in_cache("", None)

    cache_mock.query.return_value = [("SPAM#2#5#32", "eggs")]
    data = tile_cache_test.query()
    cache_mock.query.assert_called_with()
    assert data == [(("SPAM", 2, 5, 32), "eggs")]

    tile_cache_test.remove(("SPAM", 6, 3, 12))
    cache_mock.remove.assert_called_with("SPAM#6#3#12")
    
@pytest.fixture
def dumpdir():
    name = "test_dump_dir"
    try:
        shutil.rmtree(name)
    except:
        pass
    try:
        os.mkdir(name)
    except:
        pass
    yield name
    try:
        shutil.rmtree(name)
    except:
        pass

def test_Cache_dump(dumpdir):
    cache_mock = mock.Mock()
    tile_cache_test = tiles.Cache(cache_mock)
    cache_mock.query.return_value = [
        ("ONE#1#2#5", None),
        ("ONE#1#3#5", None),
        ("ONE#4#5#6", None),
        ("TWO#1#2#5", None)
        ]
    def get_data(key):
        if key == "ONE#1#2#5":
            return (b"\x89PNG", None)
        elif key == "ONE#1#3#5":
            return (b"123456JFIF", None)
        elif key == "ONE#4#5#6":
            return (b"asdfhgsdgjsdhjkg", None)
        elif key == "TWO#1#2#5":
            return (b"asdfhgsdgjsdhjkg", None)
        else:
            raise AssertionError()
    cache_mock.get_from_cache.side_effect = get_data
    tile_cache_test.dump(dumpdir)
    assert set( os.listdir(dumpdir) ) == {"ONE", "TWO"}
    assert set( os.listdir(os.path.join(dumpdir, "ONE")) ) == {"5", "6"}
    assert set( os.listdir(os.path.join(dumpdir, "ONE", "5")) ) == {"1_2.png", "1_3.jpg"}
    assert set( os.listdir(os.path.join(dumpdir, "ONE", "6")) ) == {"4_5"}
    assert set( os.listdir(os.path.join(dumpdir, "TWO")) ) == {"5"}
    assert set( os.listdir(os.path.join(dumpdir, "TWO", "5")) ) == {"1_2"}

def test_Cache_dump_must_be_empty(dumpdir):
    cache_mock = mock.Mock()
    tile_cache_test = tiles.Cache(cache_mock)

    with open(os.path.join(dumpdir, "matt.txt"), "w") as f:
        f.write("Nope")

    with pytest.raises(Exception):
        tile_cache_test.dump(dumpdir)
    
def test_Cache_clean():
    cache_mock = mock.Mock()
    tile_cache_test = tiles.Cache(cache_mock)

    cache_mock.query.return_value = [
        ("ONE#1#2#5", datetime.datetime(2017,5,6)),
        ("ONE#1#3#5", datetime.datetime(2017,5,5)),
        ("ONE#4#5#6", datetime.datetime(2017,5,4,12,30)),
        ("TWO#1#2#5", datetime.datetime(2017,5,4))
        ]
    tile_cache_test.clean(datetime.datetime(2017,5,4,12,30,1))

    cache_mock.remove.call_args_list == [
        mock.call("ONE#4#5#6"), mock.call("TWO#1#2#5") ]


tests/test_utils.py
import pytest
import unittest.mock as mock

import tilemapbase.utils as utils
import PIL.Image
import threading
import time

def test_logging():
    with mock.patch("sys.__stdout__") as stdmock:
        utils.start_logging()
        import logging
        logger = logging.getLogger("tilemapbase")
        logger.debug("Fish")

        call = stdmock.write.call_args_list[0]
        written_line = call[0][0].strip()
        assert written_line.endswith("DEBUG tilemapbase - Fish")

def test_Cache():
    c = utils.Cache()

    c[5] = "Spam"
    assert 5 in c
    assert c[5] == "Spam"
    assert 6 not in c

def test_Cache_evict():
    c = utils.Cache(2)
    c[5] = "a"
    c[6] = "b"
    c[7] = "c"
    assert set(c.keys()) == {6,7}
    assert c[6] == "b"
    assert c[7] == "c"

    c[8] = "d"
    assert set(c.keys()) == {7,8}
    assert c[7] == "c"

    c[9] = "e"
    assert set(c.keys()) == {7,9}

@pytest.fixture
def random_image():
    image = PIL.Image.new("RGB", (200, 100))
    import random
    b = bytes([random.randint(0, 255) for _ in range(200*100*3)])
    image.frombytes(b)
    return image

def test_ImageCache(random_image):
    c = utils.ImageCache()
    c[5] = random_image
    c[7] = "spam"

    assert c[7] == "spam"
    image = c[5]
    assert image.mode == random_image.mode
    assert image.size == random_image.size
    assert image.tobytes() == random_image.tobytes()

@pytest.fixture
def random_pal_image():
    image = PIL.Image.new("P", (200, 100))
    import random
    b = bytes([random.randint(0, 255) for _ in range(200*100*3)])
    image.frombytes(b)
    image.putpalette([random.randint(0, 255) for _ in range(3 * 256)])
    return image

def test_ImageCache(random_pal_image):
    c = utils.ImageCache()
    c[5] = random_pal_image
    c[7] = "spam"

    assert c[7] == "spam"
    image = c[5]
    assert image.mode == random_pal_image.mode
    assert image.size == random_pal_image.size
    assert image.tobytes() == random_pal_image.tobytes()
    assert image.getpalette() == random_pal_image.getpalette()

def test_PerThreadProvider():
    count = 0
    def factory():
        nonlocal count
        count += 1
        return count
    ptp = utils.PerThreadProvider(factory)
    assert ptp.get() == 1
    assert ptp.get() == 1

    barrier = threading.Barrier(2)
    def test():
        assert ptp.get() == 2
        barrier.wait()
    threading.Thread(target=test).start()
    assert len(ptp.active_objects()) == 2
    barrier.wait()

    called = False
    def dest(a):
        assert a == 2
        nonlocal called
        called = True
    ptp.set_destructor(dest)

    time.sleep(0.5)
    assert ptp.get() == 1
    assert len(ptp.active_objects()) == 1
    assert called


tilemapbase/cache.py
"""
cache
~~~~~

Provides a base class for caching objects which are associated with a "request"
object.  Concrete implementation using a sqlite database.

A "request" object is simply any object such that `str(obj)` returns a unique
identifier of the request.  Any object which can be converted to `bytes` by
`bytes(obj)` can be cached (and will be returned as a `bytes` object).
"""

import datetime as _datetime
import sqlite3 as _sqlite3
import os as _os
from . import utils as _utils

class Executor():
    """A base class for executing a request, if the cache does not already have
    access to the object requested.  Any class with this interface will work.
    """

    def fetch(self, request):
        """Attempt to fetch the request, returning an object (typically a
        :class:`bytes` instance, or object which can be converted to
        :class:`bytes`).  May return `None` or raise an exception on failure.
        """
        raise NotImplementedError()


class ConcreteCache():
    """Abstract base class for a class which implements storing and retrieving
    objects.  Each object should be timestamped with its last update time.
    """
    def get_from_cache(self, str_request):
        """Return `None` on failure to find.
        
        :return: Pair of (object, last_update_time)
        """
        raise NotImplementedError()

    def place_in_cache(self, str_request, obj_as_bytes):
        """Write the object to the cache.  Should use the current time when
        storing the "last update time"."""
        raise NotImplementedError()

    def query(self):
        """List all `str_request` objects which are in the cache.

        :return: List of pairs `(str_request, last_update_time)`
        """
        raise NotImplementedError()

    def remove(self, str_request):
        """Remove the item from the cache."""
        raise NotImplementedError()


class Cache(Executor):
    """A base class for a "cache".  Implements the business logic, but defers
    to two members for storing/retrieving objects, and fetching new objects.

    :param executor: The :class:`Executor` instance to delegate to if the cache
      cannot satisfy the request.
    :param cache: The :class:`ConcereteCache` instance to use for storing
      objects.
    """
    def __init__(self, executor, cache):
        self._executor = executor
        self._cache = cache
        self._expire_time = None
        pass

    def no_timeout(self):
        """Set so that no cache objects expire."""
        self._expire_time = None

    @property
    def expire_time(self):
        """The duration after which cached objects will "expire" (be removed
        from the cache).   `None` indicates no time-out.
        """
        return self._expire_time

    @expire_time.setter
    def expire_time(self, duration):
        self._expire_time = duration

    def fetch(self, request):
        str_request = str(request)

        cache = self._cache.get_from_cache(str_request)
        if self.expire_time is not None and cache is not None:
            since_refresh = _datetime.datetime.now() - cache[1]
            if since_refresh > self.expire_time:
                cache = None

        if cache is None:
            obj = self._executor.fetch(request)
            if obj is not None:
                self._cache.place_in_cache(str_request, bytes(obj))
        else:
            obj = cache[0]
        return obj


def database_exists(db_filename):
    """Attempt to open the file as a SQLite database and see if there is a
    table "cache".  Returns `True` if there is (with any schema) and `False`
    for any error.
    """
    try:
        _os.stat(db_filename)
        conn = _sqlite3.connect(db_filename)
        try:
            tables = conn.execute("SELECT name FROM sqlite_master WHERE type='table'").fetchall()
            tables = [x[0] for x in tables]
            return "cache" in tables
        except Exception:
            return False
        finally:
            conn.close()
    except Exception:
        return False


class SQLiteCache(ConcreteCache):
    """Uses a SQLite database to implement a cache.  There is one table,
    "cache", and objects are stored by the string of their "request".

    :param db_filename: The filename of the database.  If the database exists,
      we check for the existance of a table "cache", and create it if it
      doesn't exist.
    """
    def __init__(self, db_filename):
        if not database_exists(db_filename):
            self._make_database(db_filename)
        self._filename = db_filename
        self._connection_provider = _utils.PerThreadProvider(self._new)

    def _new(self):
        return _sqlite3.connect(self._filename)

    def _make_database(self, db_filename):
        conn = _sqlite3.connect(db_filename)
        try:
            conn.execute("CREATE table cache (request STRING UNIQUE, data BLOB, create_time STRING)")
        finally:
            conn.close()

    _ISO_FORMAT = "%Y-%m-%dT%H:%M:%S"

    def get_from_cache(self, str_request):
        conn = self._connection_provider.get()
        row = conn.execute("SELECT data, create_time FROM cache WHERE request=?", (str_request,)).fetchone()
        if row is None:
            return None
        update_time = _datetime.datetime.strptime(row[1], self._ISO_FORMAT)
        return row[0], update_time

    def place_in_cache(self, str_request, obj_as_bytes):
        update_time = _datetime.datetime.strftime(_datetime.datetime.now(), self._ISO_FORMAT)
        data = (str_request, obj_as_bytes, update_time)
        with self._connection_provider.get() as conn:
            conn.execute("INSERT OR REPLACE INTO cache(request, data, create_time) VALUES (?,?,?)", data)

    def query(self):
        conn = self._connection_provider.get()
        cursor = conn.execute("SELECT request, create_time FROM cache")
        out = []
        for row in cursor:
            update_time = _datetime.datetime.strptime(row[1], self._ISO_FORMAT)
            out.append((row[0], update_time))
        return out

    def remove(self, str_request):
        with self._connection_provider.get() as conn:
            conn.execute("DELETE FROM cache WHERE request=?", (str_request,))

    def close(self):
        """Close the underlying database connection (for this thread)."""
        self._connection_provider.get().close()


tilemapbase/tiles.py
"""
tiles
~~~~~

Handles reading tiles from a HTTP source, and caching.

Please note that for OpenStreetMap (and derived) tiles:

- Data is "© OpenStreetMap contributors”, see
  http://www.openstreetmap.org/copyright
- Tile set usage is subject to constraints:
  https://operations.osmfoundation.org/policies/tiles/

For Stamen maps, please see:

- http://maps.stamen.com
- For Toner and Terrain maps, Map tiles by <a href="http://stamen.com">Stamen Design</a>,
  under <a href="http://creativecommons.org/licenses/by/3.0">CC BY 3.0</a>. Data by
  <a href="http://openstreetmap.org">OpenStreetMap</a>, under
  <a href="http://www.openstreetmap.org/copyright">ODbL</a>.
- For Watercolor: Map tiles by <a href="http://stamen.com">Stamen Design</a>, under
  <a href="http://creativecommons.org/licenses/by/3.0">CC BY 3.0</a>. Data by
  <a href="http://openstreetmap.org">OpenStreetMap</a>, under
  <a href="http://creativecommons.org/licenses/by-sa/3.0">CC BY SA</a>.

For Carto maps, please see:

- https://carto.com/location-data-services/basemaps/
- Free for non-commerical use only

For other providers, copyright and usage rights may vary.
"""

from . import cache as _cache
import os as _os
import io as _io
import requests as _requests
import PIL.Image as _Image
import logging as _logging
import datetime as _datetime

# Singleton
_sqcache = None

def init(cache_filename = None, create = False):
    """Initialise the cache.  To avoid spamming the tile server, we cache the
    resulting tiles.  We are a little paranoid, so by default, the package will
    not work if no cache file exists.  Fix this by running

        init(create=True)

    to build the cache.  Optionally set a filename.
    """
    global _sqcache
    if _sqcache is not None:
        return

    if cache_filename is None:
        home = _os.path.expanduser("~")
        cache_filename = _os.path.join(home, "tilemapbase_cache.db")

    dbexists = _cache.database_exists(cache_filename) 
    if not dbexists and not create:
        msg = """The database cache file {} does no exist and no request made """\
            """to create.\n"""\
            """This could be caused by:\n"""\
            """- Running the package for the first time.  To be paranoid, we do """\
            """not automatically create a new cache.  Run `init(create=True)` """\
            """to initise the cache.\n"""\
            """- You have run the package before, but the cache has been lost. """\
            """Check that the filename is correct."""
        raise Exception(msg.format(cache_filename))

    _sqcache = _cache.SQLiteCache(cache_filename)

def close():
    """Shutdown the cache.  Call to close the connection to the database
    file."""
    global _sqcache
    if _sqcache is not None:
        try:
            _sqcache.close()
        finally:
            _sqcache = None

def _get_cache():
    global _sqcache
    if _sqcache is None:
        init()
    return _sqcache


class Cache(_cache.ConcreteCache):
    """A friendly wrapper around our cache, which decodes the key names.
    For this class, "request strings" are tuples of the form
    `(name, x, y, zoom)`.
    """
    def __init__(self, delegate):
        self._delegate = delegate

    @staticmethod
    def make_request_string(name, x, y, zoom):
        return "{}#{}#{}#{}".format(name, x, y, zoom)

    @staticmethod
    def split_request_string(str_request):
        name, x, y, zoom = str_request.split("#")
        return name, int(x), int(y), int(zoom)

    def get_from_cache(self, key):
        return self._delegate.get_from_cache(self.make_request_string(*key))

    def place_in_cache(self, str_request, obj_as_bytes):
        raise NotImplementedError()

    def query(self):
        return [(self.split_request_string(sr), ti) 
            for (sr, ti) in self._delegate.query()]

    def remove(self, key):
        self._delegate.remove(self.make_request_string(*key))

    def dump(self, dirname):
        """Dump all the files to the given directory.  The directory should
        be empty.  The directory structure will be:
           - dirname
             - Name of tile provider
               - Zoom level
                 - Files in format "x_y.png" or "x_y.jpg" or "x_y"

        We attempt to auto detect PNG and JPG files.
        """
        if _os.listdir(dirname) != []:
            raise Exception("Directory needs to be empty")
        
        for (key, _) in self.query():
            name, x, y, zoom = key
            filename = _os.path.join(dirname, name)
            try:
                _os.mkdir(filename)
            except:
                pass
            filename = _os.path.join(filename, str(zoom))                
            try:
                _os.mkdir(filename)
            except:
                pass
            data, _ = self.get_from_cache(key)
            if data[:4] == b"\x89PNG":
                ext = ".png"
            elif data[6:10] == b"JFIF":
                ext = ".jpg"
            else:
                ext = ""
            filename = _os.path.join(filename, "{}_{}{}".format(x, y, ext))
            with open(filename, "wb") as f:
                f.write(data)

    def clean(self, cutoff):
        """Remove all files from the cache which were written before the
        cutoff.

        :param cutoff: Datetime
        """
        query = list(self.query())
        for (key, time) in query:
            if time < cutoff:
                self.remove(key)


def get_cache():
    """Get access to the underlying DB cache.  For advanced usage only."""
    return Cache(_get_cache())


class _TilesExecutor(_cache.Executor):
    """Private class to run the HTTP request."""
    def __init__(self, parent):
        self.parent = parent
        self.logger = _logging.getLogger(__name__)

    def fetch(self, request):
        url = self.parent._request_http(request)
        self.logger.info("Requesting %s", url)
        # TODO: I didn't understand the default value of headers.
        #       So not to rely on any, avoid setting it if unset.
        if self.parent.headers is not None:
            response = _requests.get(url, headers=self.parent.headers)
        else:
            response = _requests.get(url)
        if not response.ok:
            raise IOError("Failed to download {}.  Got {}".format(url, response))
        try:
            _Image.open(_io.BytesIO(response.content))
        except:
            raise IOError("Received invalid tile from {}.".format(url), response)
        return response.content


class Tiles():
    """Class to fetch a tile as an image; transparently handles caching issues.
    Tiles will be expired from the cache after 2 months.
    
    :param request_string: A string which when used with `format` will give
      a well-formed URL for the tile.  For example, for standard OSM this is
      "http://a.tile.openstreetmap.org/{zoom}/{x}/{y}.png".
    :param source_name: A short, human-readable name, e.g. "OSM".  Should be
      unique, as also used by the cache.
    :param tilesize: The size of the (square) tiles, defaults to 256.
    :param maxzoom: The maximum tile zoom level, defaults to 19.
    :param headers: Dictionary of http headers for server requests
    """
    def __init__(self, request_string, source_name, tilesize=256, maxzoom = 19, headers = None):
        self.request = request_string
        self.name = source_name
        self.headers = headers
        self._maxzoom = maxzoom
        self._tilesize = tilesize
        self._cache = None

    def get_tile(self, x, y, zoom):
        """Attempt to fetch the tile at the specified coords and zoom level.

        :param x: X coord of the tile; must be between 0 (inclusive) and
          `2**zoom` (exclusive).
        :param y: Y coord of the tile.
        :param zoom: Integer, greater than or equal to 0.  19 is the commonly
          supported maximum zoom.

        :return: `None` for (cache related) failure, or a :package:`Pillow`
          image object of the tile.
        """
        tile = self._get_cache().fetch(self._request_string(x, y, zoom))
        if tile is None:
            return None
        try:
            fp = _io.BytesIO(tile)
            return _Image.open(fp)
        except:
            raise RuntimeError("Failed to decode data for {} - {}x{} @ {} zoom".format(self.name, x, y, zoom))

    @property
    def maxzoom(self):
        """The maximum zoom level supported by this tile provider."""
        return self._maxzoom

    @property
    def tilesize(self):
        """The size of the tiles."""
        return self._tilesize

    def _request_string(self, x, y, zoom):
        """Encodes the tile coords, zoom, and name into a string for the
        database."""
        return Cache.make_request_string(self.name, x, y, zoom)

    def _request_http(self, request_string):
        parts = request_string.split("#")
        if parts[0] != self.name:
            raise ValueError("Build for '{}' but asked to decode '{}'".format(self.name, parts[0]))
        x, y, zoom = [int(t) for t in parts[1:]]
        return self.request.format(x=x, y=y, zoom=zoom)

    # Lazy initialisation
    def _get_cache(self):
        if self._cache is None:
            dbcache = _get_cache()
            executor = _TilesExecutor(self)
            self._cache = _cache.Cache(executor, dbcache)
            self._cache.expire_time = _datetime.timedelta(days=60)
        return self._cache


"""We have to have functions for OSM realted requests so users can pass their own
user agent strings
"""
def build_OSM(headers={"User-Agent":"TileMapBase"}):
    """Standard Open Street Map tile server."""
    OSM = Tiles("http://a.tile.openstreetmap.org/{zoom}/{x}/{y}.png", "OSM", headers=headers)
    return OSM

def build_OSM_Humanitarian(headers={"User-Agent":"TileMapBase"}):
    """Humanitarian Open Street Map tile server."""
    OSM_Humanitarian = Tiles("http://a.tile.openstreetmap.fr/hot/{zoom}/{x}/{y}.png ", "OSM_HUMANITARIAN", headers=headers)
    return OSM_Humanitarian

"""Stamen, Toner, Standard."""
Stamen_Toner = Tiles("http://tile.stamen.com/toner/{zoom}/{x}/{y}.png", "STAMEN_TONER")

"""Stamen, Toner, Hybrid."""
Stamen_Toner_Hybrid = Tiles("http://tile.stamen.com/toner-hybrid/{zoom}/{x}/{y}.png", "STAMEN_TONER_Hybrid")

"""Stamen, Toner, Labels."""
Stamen_Toner_Labels = Tiles("http://tile.stamen.com/toner-labels/{zoom}/{x}/{y}.png", "STAMEN_TONER_Labels")

"""Stamen, Toner, Lines."""
Stamen_Toner_Lines = Tiles("http://tile.stamen.com/toner-lines/{zoom}/{x}/{y}.png", "STAMEN_TONER_Lines")

"""Stamen, Toner, Background."""
Stamen_Toner_Background = Tiles("http://tile.stamen.com/toner-background/{zoom}/{x}/{y}.png", "STAMEN_TONER_Background")

"""Stamen, Toner, Lite."""
Stamen_Toner_Lite = Tiles("http://tile.stamen.com/toner-lite/{zoom}/{x}/{y}.png", "STAMEN_TONER_Lite")

"""Stamen, Terrain."""
Stamen_Terrain = Tiles("http://tile.stamen.com/terrain/{zoom}/{x}/{y}.jpg", "STAMEN_TERRAIN")

"""Stamen, Terrain, Labels"""
Stamen_Terrain_Labels = Tiles("http://tile.stamen.com/terrain-labels/{zoom}/{x}/{y}.jpg", "STAMEN_TERRAIN_Labels")

"""Stamen, Terrain, Lines"""
Stamen_Terrain_Lines = Tiles("http://tile.stamen.com/terrain-lines/{zoom}/{x}/{y}.jpg", "STAMEN_TERRAIN_Lines")

"""Stamen, Terrain, Background"""
Stamen_Terrain_Background = Tiles("http://tile.stamen.com/terrain-background/{zoom}/{x}/{y}.jpg", "STAMEN_TERRAIN_Background")

"""Stamen, Watercolour"""
Stamen_Watercolour = Tiles("http://tile.stamen.com/watercolor/{zoom}/{x}/{y}.jpg", "STAMEN_WATERCOLOUR")

"""Carto, Light"""
Carto_Light = Tiles("http://a.basemaps.cartocdn.com/light_all/{zoom}/{x}/{y}.png", "CARTO_LIGHT")

"""Carto, Light, Labels"""
Carto_Light_Labels = Tiles("http://a.basemaps.cartocdn.com/light_only_labels/{zoom}/{x}/{y}.png", "CARTO_LIGHT_LABELS")

"""Carto, Light, No labels"""
Carto_Light_No_Labels = Tiles("http://a.basemaps.cartocdn.com/light_nolabels/{zoom}/{x}/{y}.png", "CARTO_LIGHT_NOLABELS")

"""Carto, Dark"""
Carto_Dark = Tiles("http://a.basemaps.cartocdn.com/dark_all/{zoom}/{x}/{y}.png", "CARTO_DARK")

"""Carto, Dark, Labels"""
Carto_Dark_Labels = Tiles("http://a.basemaps.cartocdn.com/dark_only_labels/{zoom}/{x}/{y}.png", "CARTO_DARK_LABELS")

"""Carto, Dark, No labels"""
Carto_Dark_No_Labels = Tiles("http://a.basemaps.cartocdn.com/dark_nolabels/{zoom}/{x}/{y}.png", "CARTO_DARK_NOLABELS")
